presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/SpatialPartitioningInternalAggregateFunction.java

```diff
@@ -0,0 +1,108 @@
+import com.facebook.presto.spi.type.StandardTypes;
+import io.airlift.json.JsonCodec;
+import io.airlift.json.JsonCodecFactory;
+import io.airlift.slice.Slice;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static com.facebook.presto.geospatial.KDBTree.buildKDBTree;
+import static com.facebook.presto.geospatial.serde.GeometrySerde.deserializeEnvelope;
+import static com.facebook.presto.plugin.geospatial.GeometryType.GEOMETRY_TYPE_NAME;
+import static com.facebook.presto.spi.type.StandardTypes.INTEGER;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static java.lang.Math.floor;
+import static java.lang.Math.toIntExact;
+
+@AggregationFunction(value = "spatial_partitioning", decomposable = false, isOrderSensitive = true, hidden = true)
+public class SpatialPartitioningInternalAggregateFunction
+{
+    private static final JsonCodec<KDBTree> KDB_TREE_CODEC = new JsonCodecFactory().jsonCodec(KDBTree.class);
+    private static final int MAX_SAMPLE_COUNT = 1_000_000;
+
+    private SpatialPartitioningInternalAggregateFunction() {}
+
+    @InputFunction
+    public static void input(SpatialPartitioningState state, @SqlType(GEOMETRY_TYPE_NAME) Slice slice, @SqlType(INTEGER) long partitionCount)
+    {
+        Envelope envelope = deserializeEnvelope(slice);
+        if (envelope == null) {
+            return;
+        }
+
+        Rectangle extent = new Rectangle(envelope.getXMin(), envelope.getYMin(), envelope.getXMax(), envelope.getYMax());
+
+        if (state.getCount() == 0) {
+            state.setPartitionCount(toIntExact(partitionCount));
+            state.setExtent(extent);
+            state.setSamples(new ArrayList<>());
+        }
+        else {
+            state.setExtent(state.getExtent().merge(extent));
+        }
+
+        List<Rectangle> samples = state.getSamples();
+        if (samples.size() <= MAX_SAMPLE_COUNT) {
+            samples.add(extent);
+        }
+        else {
+            int sampleIndex = (int) floor(ThreadLocalRandom.current().nextDouble() * state.getCount());
+            if (sampleIndex < MAX_SAMPLE_COUNT) {
+                samples.set(sampleIndex, extent);
+            }
+        }
+
+        state.setCount(state.getCount() + 1);
+    }
+
+    @CombineFunction
+    public static void combine(SpatialPartitioningState state, SpatialPartitioningState otherState)
+    {
+        throw new UnsupportedOperationException("spatial_partitioning must run on a single node");
+    }
+
+    @OutputFunction(StandardTypes.VARCHAR)
+    public static void output(SpatialPartitioningState state, BlockBuilder out)
+    {
+        if (state.getCount() == 0) {
+            out.appendNull();
+            return;
+        }
+
+        List<Rectangle> samples = state.getSamples();
+
+        int partitionCount = state.getPartitionCount();
+        int maxItemsPerNode = (int) (Math.ceil(samples.size() * 1.0 / partitionCount));
```

Comment:
`toIntExact`

Reply:
@arhimondr `toIntExact` doesn't take `double`

Reply:
Sorry, i meant `int maxItemsPerNode = toIntExact((Math.ceil(samples.size() * 1.0 / partitionCount)))`

Reply:
Oh, sorry. I didn't know `Math.ceil` returns `double`. That is lame. Please use `roundToInt(sample.size() * 1.0 / partitionCount, CEILING)`. The same in the other place where you use `ceil`. It is overflow and `NaN` safe.

Reply:
@arhimondr Thank you. Will do.